Fix Part of #11496: Lint Check for filter Functions

ruinan-liu opened the pull request:
## Overview

1. This PR fixes or fixes part of #11496.
2. This PR does the following: Added lint checks to ensure no `.filter()` function is used and instead should use `for` loop. 

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

No user-facing interface. To confirm that the test currently passes, see the CI checks on this PR. I confirmed locally that the test fails when it should.

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Commit:
Added customer linter check to ensure no filter() function is used. Replaced Filter function with for loop for files in core/tests/protractor_utils/ directory

Commit:
Fixed linting problem during push

Commit:
Fixing linting issues

Commit:
Merge branch 'develop' of https://github.com/ruinan-liu/oppia into develop
Merging to fix the Pull error

Commit:
Merge remote-tracking branch 'upstream/develop' into develop
Merging for the newest comits to the upstream repo

U8NWXD commented:
@ruinan-liu please change the title and description to note that this only fixes _part_ of #11496

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into develop
Mereging the get the latest functionality

ruinan-liu committed:
Merge branch 'oppia:develop' into develop

Commit:
Merge branch 'develop' of https://github.com/ruinan-liu/oppia into develop
Obtaining the latest build

U8NWXD commented on scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js:
```diff
@@ -130,6 +130,15 @@ ruleTester.run('protractor-practices', rule, {
         message: 'Please do not use .then(), consider async/await instead',
       }],
     },
+    {
+      code:
+      `genderList.filter(function() {
+        gender = 'female';
+      })`,
```
Can you make these test cases more comprehensive? For example:

* A comment that includes "filter" should pass
* Reading the `filter` attribute of an object (but not calling it as a function) should pass
* Passing a function to `filter()` that was defined elsewhere should fail (e.g. `var func = function() {...} ... elems.filter(myFunc)`

U8NWXD commented on core/tests/protractor_utils/CreatorDashboardPage.js:
```diff
@@ -69,14 +69,17 @@ var CreatorDashboardPage = function() {
-        text === explorationTitle
-      );
-    });
+      if (text.startsWith(explorationTitle + '\n' ||
+      text === explorationTitle)) {
```
indentation here is hard to read, and I think the parentheses are wrong. How about this instead:

```suggestion
      if (
        text.startsWith(explorationTitle + '\n') || text === explorationTitle
      ) {
```

U8NWXD commented on core/tests/protractor_utils/LibraryPage.js:
```diff
@@ -61,13 +61,16 @@ var LibraryPage = function() {
-        var tileTitle = await tile.element(
-          expSummaryTileTitleLocator).getText();
-        return (tileTitle === name);
+    // Using for loop instead of .filter() function.
```
Don't need this comment. ditto elsewhere

U8NWXD commented on core/tests/protractor_utils/general.js:
```diff
@@ -70,9 +70,16 @@ var checkForConsoleErrors = async function(
+  var browserErrors = [];
+  for (i = 0; i < browserLogs.length; i++) {
+    var logEntry = browserLogs[i];
+    if (logEntry.level.value > CONSOLE_LOG_THRESHOLD &&
+      errorsToIgnore.every(e => logEntry.message.match(e) === null)) {
+      browserErrors.push(logEntry);
```
```suggestion
    if (logEntry.level.value > CONSOLE_LOG_THRESHOLD &&
      errorsToIgnore.every(e => logEntry.message.match(e) === null)
    ) {
      browserErrors.push(logEntry);
```

U8NWXD commented:
Thanks for your PR @ruinan-liu! I left some comments

ruinan-liu committed:
Update core/tests/protractor_utils/CreatorDashboardPage.js

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

ruinan-liu committed:
Update core/tests/protractor_utils/general.js

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

Commit:
Added more passing test as well as failing test, also deleted unnecessary comment

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into add-lint-checks
Merging for push to add-linter-checks branch

Commit:
Merge branch 'add-lint-checks' of https://github.com/ruinan-liu/oppia into add-lint-checks
Merging for push

ruinan-liu commented:
@U8NWXD Just updated the PR, Thanks!

ruinan-liu committed:
Merge branch 'oppia:develop' into develop

U8NWXD commented on scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js:
```diff
@@ -67,6 +67,12 @@ ruleTester.run('protractor-practices', rule, {
+  }, {
+    code: '/* Filter results from array*/'
+  }, {
+    code: 'var filtered_fruti = apple.filter'
```
```suggestion
    code: 'var filtered_fruit = apple.filter'
```

U8NWXD commented on scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js:
```diff
@@ -134,7 +140,18 @@ ruleTester.run('protractor-practices', rule, {
+      function gradeList(name, gradebook){
+        return gradebook[name]
+      }
+      gradeList.filter(gradeList(name, gradebook));`,
```
I think this is a more realistic use of `filter`:

```suggestion
      gradeList.filter((name) => gradeList(name, gradebook));`,
```

U8NWXD commented on scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js:
```diff
@@ -134,7 +140,18 @@ ruleTester.run('protractor-practices', rule, {
-      })`,
+      });`,
+      errors: [{
+        message: 'Please do not use .filter(), consider using a for loop',
```
nit: since "consider using a for loop" is a full independent clause, you need a semicolon instead of a comma

```suggestion
        message: 'Please do not use .filter(); consider using a for loop',
```

U8NWXD commented on scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js:
```diff
@@ -67,6 +67,12 @@ ruleTester.run('protractor-practices', rule, {
   }, {
     code: 'var modal = element(by.css("option:checked"));'
+  }, {
+    code: '// Filter results from the array'
```
To make clearer that we're testing the lint rule:

```suggestion
    code: '// Filter results from the array using arr.filter()'
```

U8NWXD commented on scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js:
```diff
@@ -67,6 +67,12 @@ ruleTester.run('protractor-practices', rule, {
+  }, {
+    code: '// Filter results from the array'
+  }, {
+    code: '/* Filter results from array*/'
```
ditto:

```suggestion
    code: '/* Filter results from array using arr.filter() */'
```

U8NWXD reviewed (changes requested):
Just a few more comments

ruinan-liu committed:
Update scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js

Done

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

ruinan-liu committed:
Update scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js

Done

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

ruinan-liu committed:
Update scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js

Done

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

ruinan-liu committed:
Update scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js

Done

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

ruinan-liu committed:
Update scripts/linters/custom_eslint_checks/rules/protractor-practices.spec.js

Done

Co-authored-by: U8N WXD <U8NWXD@users.noreply.github.com>

Commit:
Merge branch 'develop' of https://github.com/ruinan-liu/oppia into add-lint-checks
Merging to resolve a conflict

Commit:
Adding most current build for PR13849

Commit:
Fix linting error

Commit:
Fix merging conflicts

ruinan-liu commented:
@U8NWXD Thanks, I just updated the PR with your suggestion as well as the latest build. 

Also, some of the frontend tests shows fails, however does that relate to my code since it passed all test locally?

Thanks

U8NWXD reviewed (approved):
This looks great @ruinan-liu! For future reference, please don't resolve comments left by reviewers. Here's how we explain why on the [wiki](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve):

> Also, please do not mark the comment as resolved, since this just makes it harder to actually read the comment thread.

U8NWXD commented:
@DubeySandeep @Hudda PTAL

ruinan-liu commented:
@U8NWXD Would you mind taking a look and merging the PR? Thanks, I was worried that it we take too long it might be stale build again.

U8NWXD commented:
@ruinan-liu it looks like [one of your tests](https://github.com/oppia/oppia/pull/13849/checks?check_run_id=3658505680) needs to be updated:

```
       AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'Please do not use .filter(), consider using a for loop'
- 'Please do not use .filter(); consider using a for loop'
```

Commit:
Merge branch 'develop' of https://github.com/oppia/oppia into add-lint-checks
Fetching to obtain the latest

Commit:
Fixed checking error

ruinan-liu commented:
@U8NWXD Apologies, just updated the and ran the changes locally, all test passed. 

Thanks!

U8NWXD commented:
I'm seeing a suspicious number of E2E test failures that look like this: `Failed: Could not find exploration tile with name ...`. Could you dig into them a little? I wonder if something's wrong with the changes to `core/tests/protractor_utils/LibraryPage.js`?

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

ruinan-liu commented:
@U8NWXD I took a quick look regarding my 5 failed e2e tests and the error seems to stem from 2 files `LibraryPage.js` and `CreatorDashboardPage.js`. It seems that the error is coming from the function with name `_getExplorationElements` in `CreatorDashboard.js` and `_getExplorationElements` in `LibraryPage.js`.

Both functions I modified because of the need to remove filter function. However, I can not reproduce the error locally, I noticed that in `libraryFlow.js` file, it calls the function `_getExplorationElements` so I run the e2e test command `python -m scripts.run_e2e_tests --suite="library"` however no error is surfaced. 

I also tried to run the exact command in e2e test `python -m scripts.run_e2e_tests --suite="fileUploadFeatures" --prod_env` and still I could not observe the error on my local environment

Do you know how can I reproduce the error locally? 

Thanks!

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

Commit:
Fixing e2e run error

Commit:
Fixing logic for filter function

Commit:
Fixed logic for filter function in ProfilePage.js and general.js

ruinan-liu commented:
@U8NWXD Although I still can not test locally, I think I have found the error in my code. For example, in `CreatorDashboard.js` I did not append the actually element of the array `allExplorationCards[i]` to the filtered list, rather I pushed the `text` to the array. 

I just fixed all files and we could run the test to see whether it resolves

Thanks

U8NWXD commented:
@ruinan-liu great! I've started your tests. Please assign me once they finish

ruinan-liu commented:
@U8NWXD I saw that still 6 tests fails, and mostly it is around changing the filter function. Would you mind taking a look at `CreatorDashboard.js` and files changed and unstuck me a little? 

All I changed is that I replaced the `.filter()` function with a for loop, and I still used the conditional statement in the filter function to decide which element gets pushed to the `filteredarray` and then simply returning the array. 

Would you mind shedding some light into this?

Thanks

U8NWXD commented:
I have two suggestions:

1. The code you changed includes `var text = await allExplorationCards[i].getText();`. Maybe try using `action.getText(...)` instead? That will wait for the exploration card to load before trying to read its text, which might help if the `for` loop runs faster than `filter`.
2. You could debug using `console.log` statements inside your loop. Maybe try printing out exploration titles before and after your changes to see if they differ.

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

Commit:
changed from await to action.getText() in case for loop is faster causing empty filtered list

U8NWXD commented on core/tests/protractor_utils/CreatorDashboardPage.js:
```diff
@@ -69,14 +69,17 @@ var CreatorDashboardPage = function() {
-      var text = await tile.getText();
+    var filteredExplorationCards = [];
+    for (i = 0; i < allExplorationCards.length; i++) {
+      var text = action.getText(allExplorationCards[i]);
```
I don't think this is a correct usage of `action.getText`. Here's the code:

https://github.com/oppia/oppia/blob/5d81029defe519017194864bf8a0586fdb511f3d/core/tests/protractor_utils/action.js#L54

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

ruinan-liu commented:
@U8NWXD Thanks for the suggestions, I figured out how to test end to end locally and am still fixing the issue. 

It is particular hard for me because originally we have `async` and `await` in the filter function and right now the for loops runs faster than the await. 

Right now, I am able to get the text from the tile in `CreatorDashboard.js` and is figuring out how to iterate over the `ElementArrayFinder` object for `allExplorationCards.` Currently I am trying out various suggestion online including using `.then(...)`. Below is a code block that I am trying to replace `.filter()` function.

However, the issue is that it seems to only go through the loop once, even though `tile.length = 3`, it will not go through the loop for `i=1,2` thus resulting an empty `filteredExplorationCards`. After figuring out and letting the array iterate through for `3` iterations, I think the issue will get resolved. 

Apologies for taking a little while to fix this and thanks for your help!

![Screen Shot 2021-10-04 at 10 44 05 PM](https://user-images.githubusercontent.com/72321026/135956716-9fc94772-aac4-45e5-8e7f-1d345fed2e3a.png)

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

U8NWXD commented:
@ruinan-liu I heard you had some questions about why we want to eliminate `.filter()`. This comes from when we did the async-await migration. We found conflicting advice online about how to await a `.filter()` call (some sources said to use `Promise.all()`, but others said that caused bugs). When we were doing the migration, we found cases where `Promise.all()` caused problems, but we also found problems when we tried to directly await a `.filter()` call. Since we couldn't find a good solution, we decided to avoid `.filter()` all-together. This policy is documented in the [wiki](https://github.com/oppia/oppia/wiki/End-to-End-Tests#anti-patterns), but we hadn't added a lint rule for it yet.

U8NWXD commented:
Some notes on your screenshot:

* Based on my (admittedly limited) understanding of JS promises, you shouldn't need to use `.then()`. When you `await` something that returns a promise (btw any function with `async` returns a promise), JS should automatically wait for that promise to resolve before proceeding. That said, if `.then()` helps you debug, then go for it!
* What you're describing sounds like a control flow error where the function is returning before everything inside has completed. In my experience, these usually come from a missing `await` or a missing `async` somewhere (in this function, in one of the places where it's used, or in one of the functions it calls)

ruinan-liu commented:
@U8NWXD Thanks for the detailed explanation and for looking into it. Just want to clarify one more thing, do we just want to aviod javascript's `.filter()` function or we want to avoid **all** use of `.filter()` function. The reason is because Javascript's [function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) is a little bit different compared to Protractor's filter [function](https://www.protractortest.org/#/api?view=ElementArrayFinder.prototype.filter). 

From my understand, protractor's filter function returns a filtered ElementArrayFinder, if we use a `for` loop to replace protractor's filter function, then we will returns an filtered array of ElementFinder (The type is different).

Want to double check and make sure that we want to remove all filter function, not just the Javascript's built in Array's filter function. 

Thanks!

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

Commit:
See PR Comment for more detail

Commit:
Merge branch 'add-lint-checks' of https://github.com/ruinan-liu/oppia into add-lint-checks
Pulling the latest from oppia/oppia

Commit:
see PR for more info

ruinan-liu commented:
@U8NWXD Since protractor's `.filter()` function is part of their [api](https://www.protractortest.org/#/api?view=ElementArrayFinder.prototype.filter) it is a little bit hard to replace a for loop with it with the reason being:
1. The return data for Protractor's `.filer()` is of type `ElementArrayFinder`. If we use a for loop, then our return type is an array of `ElementFinder` which is incorrect. 
2. I did not find a way to declare an empty `elementArrayFinder`

I revert 3 files `CreatorDashboardPage.js`, `LibraryPage.js` and `ProfilePage.js` to use Protractor's `.filter()` function. I think we can impose this rule on all other DS except for Protractor since the return data type is different `Array` compared to `ElementArrayFinder`

Please let me what you thnink. 
Thanks!

U8NWXD commented:
@ruinan-liu I didn't realize there were two different `.filter()` functions. Thanks for explaining! Is there a way for the lint check to distinguish between the two? That would be required for us to allow Protractor's `.filter()` function, right?

ruinan-liu commented:
@U8NWXD I assume that if we can check the type of the Data Structure the .filter() is applying to. For example, we will only throw an error when the DS is not of type `ElementArrayFinder.` Let me see how we could implement this in linter functions

ruinan-liu commented:
@U8NWXD I have found Protarctor's Filter()'s source code. Let me see whether we can replicate or create our own filter function just for the protractor utils

i will update the PR later

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

U8NWXD commented:
@ruinan-liu 

> @U8NWXD I assume that if we can check the type of the Data Structure the .filter() is applying to. For example, we will only throw an error when the DS is not of type ElementArrayFinder. Let me see how we could implement this in linter functions

Unfortunately, I don't think this is possible. ESLint just looks at the text of the code; I don't think it knows types. I wrote up a bit of explanation about this on our [wiki page for ESLint rules](https://github.com/oppia/oppia/wiki/Custom-ESLint-checks#limits-of-linters)

> @U8NWXD I have found Protarctor's Filter()'s source code. Let me see whether we can replicate or create our own filter function just for the protractor utils
>
> i will update the PR later

One option would be to have a function that asserts that it's being given an ElementArrayFinder and then just calls `.filter()` on it. We would then add a comment to tell ESLint to ignore that one line. Here's a rough sketch of what the code could be:

```text
function myFilter(arr, func):
    assert arr is of type ElementArrayFinder
    return arr.filter(func)  // ESLint ignore
```

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

Commit:
eslint checks

Commit:
Adding lint checks

ruinan-liu commented:
@U8NWXD I tried to create a function that you suggested in the previous comment but is not very successful. However I removed the 3 files from excluded files in `.eslintc` and added `// eslint-disable-line` to the 3 `.filter` functions. 

Would this work?

Thanks!

U8NWXD commented:
> @U8NWXD I tried to create a function that you suggested in the previous comment but is not very successful. However I removed the 3 files from excluded files in .eslintc and added // eslint-disable-line to the 3 .filter functions.

@ruinan-liu Would this mean that if someone added new code that used Protractor's `.filter()` function, they would have to add `// eslint-disable-line`? If so, then I don't think this will work. We'll just end up having to disable eslint all the time.

What problems did you encounter with creating a new function?

ruinan-liu commented:
> > @U8NWXD I tried to create a function that you suggested in the previous comment but is not very successful. However I removed the 3 files from excluded files in .eslintc and added // eslint-disable-line to the 3 .filter functions.
> 
> @ruinan-liu Would this mean that if someone added new code that used Protractor's `.filter()` function, they would have to add `// eslint-disable-line`? If so, then I don't think this will work. We'll just end up having to disable eslint all the time.
> 
> What problems did you encounter with creating a new function?

@U8NWXD Do you know which files I should add this function in? Should I create a new `.js` file to hold this function? Also I don't think you can assert it is type of `ElementArrayFinder` because it will always show as object. I guess I can try to name as `element_array_filter` and in our documentation we specify that do not use this function for other instead of elementArrayFinder?

U8NWXD commented:
@ruinan-liu for the assertion, can you use [`instanceof`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof)? I think that's the canonical way to check whether an object is from a certain class.

You can add your function to `protractor_utils/general.js`

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

ruinan-liu committed:
Merge branch 'oppia:develop' into add-lint-checks

DubeySandeep commented:
@ruinan-liu Is this PR ready for review?

(Please assign reviewers once it's ready for review, in case you don't have the rights to assign reviewers, oppiaBot can help you! Please check the "PR pointers" section in the PR description!
